modules/aggs-matrix-stats/src/main/java/org/elasticsearch/search/aggregations/matrix/stats/InternalMatrixStats.java

```diff
@@ -0,0 +1,272 @@
+            ArrayList<ArrayList<Double>> cor = new ArrayList<>(fieldNames.size());
+            for (String y : fieldNames) {
+                ArrayList<Double> covRow = new ArrayList<>(fieldNames.size());
+                ArrayList<Double> corRow = new ArrayList<>(fieldNames.size());
+                for (String x : fieldNames) {
+                    covRow.add(results.getCovariance(x, y));
+                    corRow.add(results.getCorrelation(x, y));
+                }
+                cov.add(covRow);
+                cor.add(corRow);
+            }
+            builder.field(Fields.COVARIANCE, cov);
+            builder.field(Fields.CORRELATION, cor);
+        }
+
+        return builder;
+    }
+
+    @Override
+    public Object getProperty(List<String> path) {
+        if (path.isEmpty()) {
+            return this;
+        } else if (path.size() == 1) {
+            String coordinate = path.get(0);
+            if (results == null) {
+                results = MatrixStatsResults.EMPTY();
+            }
+            switch (coordinate) {
+                case "counts":
+                    return results.getFieldCounts();
+                case "means":
+                    return results.getMeans();
+                case "variances":
+                    return results.getVariances();
+                case "skewness":
+                    return results.getSkewness();
+                case "kurtosis":
+                    return results.getKurtosis();
+                case "covariance":
+                    return results.getCovariances();
+                case "correlation":
+                    return results.getCorrelations();
+                default:
+                    throw new IllegalArgumentException("Found unknown path element [" + coordinate + "] in [" + getName() + "]");
+            }
+        } else {
+            throw new IllegalArgumentException("path not supported for [" + getName() + "]: " + path);
+        }
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        // write running stats
+        if (stats == null || stats.docCount == 0) {
+            out.writeVLong(0);
+        } else {
+            out.writeVLong(stats.docCount);
+            stats.writeTo(out);
+        }
+
+        // write results
+        if (results == null || results.getDocCount() == 0) {
+            out.writeVLong(0);
+        } else {
+            out.writeVLong(results.getDocCount());
+            results.writeTo(out);
+        }
+    }
+
+    @Override
+    protected void doReadFrom(StreamInput in) throws IOException {
+        // read stats count
+        final long statsCount = in.readVLong();
+        if (statsCount > 0) {
+            stats = new RunningStats(in);
+            stats.docCount = statsCount;
+        }
+
+        // read count
+        final long count = in.readVLong();
+        if (count > 0) {
+            results = new MatrixStatsResults(in);
+        }
```

Comment:
I am not happy that if stats or results is null on one side of the wire, then it becomes an empty instance on the other side of the wire. I would like it better if we either make sure that stats/results is never null or if we transport null so that null on one side is still null on the other side (I like the former better, but the latter would work for me too if keeping nulls is simpler)